emu/src/cpu/arm/operations.rs

```diff
@@ -841,6 +836,264 @@ impl Arm7tdmi {
+        );
+
+        if set_condition_codes {
+            self.cpsr.set_carry_flag(result.1);
+            self.cpsr.set_overflow_flag(result.1);
+            self.cpsr.set_zero_flag(result.0 == 0);
+        }
+    }
+
+    pub fn umlal(&mut self, op_code: ArmModeOpcode) {
+        let set_condition_codes = op_code.get_bit(20);
+        assert!(0b1001u32 == op_code.get_bits(4..=7));
+        assert!(0u32 == op_code.get_bits(25..=27));
+
+        let rm_operand_register = op_code.get_bits(0..=3);
+        let rs_operand_register = op_code.get_bits(8..=11);
+        let rdhi_destination_register = op_code.get_bits(16..=19);
+        let rdlo_destination_register = op_code.get_bits(12..=15);
+
+        let rdhi_register_value =
+            self.registers
+                .register_at(rdhi_destination_register.try_into().unwrap()) as u64;
+        let rdlo_register_value =
+            self.registers
+                .register_at(rdlo_destination_register.try_into().unwrap()) as u64;
+        let rdhilo_register_value = rdhi_register_value << 32 | rdlo_register_value;
+        let rm_operand_value =
+            self.registers
+                .register_at(rm_operand_register.try_into().unwrap()) as u64;
+        let rs_operand_value =
+            self.registers
+                .register_at(rs_operand_register.try_into().unwrap()) as u64;
+
+        let result_mul = u64::overflowing_mul(rm_operand_value, rs_operand_value);
+        let result = u64::overflowing_add(result_mul.0, rdhilo_register_value);
+        self.registers.set_register_at(
+            rdlo_destination_register.try_into().unwrap(),
+            result.0.get_bits(0..=31).try_into().unwrap(),
+        );
+        self.registers.set_register_at(
+            rdhi_destination_register.try_into().unwrap(),
+            result.0.get_bits(32..=63).try_into().unwrap(),
+        );
+
+        if set_condition_codes {
+            let any_overflow = result_mul.1 || result.1;
+            self.cpsr.set_carry_flag(any_overflow);
+            self.cpsr.set_overflow_flag(any_overflow);
+            self.cpsr.set_zero_flag(result.0 == 0);
+        }
+    }
+
+    pub fn smull(&mut self, op_code: ArmModeOpcode) {
+        let set_condition_codes = op_code.get_bit(20);
+        assert!(0b1001u32 == op_code.get_bits(4..=7));
+        assert!(0u32 == op_code.get_bits(25..=27));
+
+        let rm_operand_register = op_code.get_bits(0..=3);
+        let rs_operand_register = op_code.get_bits(8..=11);
+        let rdhi_destination_register = op_code.get_bits(16..=19);
+        let rdlo_destination_register = op_code.get_bits(12..=15);
+
+        let rm_operand_value = self
+            .registers
+            .register_at(rm_operand_register.try_into().unwrap());
+        let rs_operand_value = self
+            .registers
+            .register_at(rs_operand_register.try_into().unwrap());
+
+        let rm_operand_value_sgn: i64 = i32::from_be_bytes(rm_operand_value.to_be_bytes()) as i64;
```

Comment:
Doesn't something like `rm_operand_value as i64` work?

Reply:
(same in similar cases)

Reply:
Thinking about this a bit more, would this line work?

I mean, will it do the sign-extension on the bits?

Imagine the output of `from_be_bytes` is a negative number (so it has a 1 as bit 31), when we do `as i64`, would this sign be extended (putting `1` to all the new 32 bits to the left)?

(There's the same problem with the `rm_operand_value as i64` I suggested before)

We have `sign_extend` for unsigned types if needed

Reply:
![image](https://github.com/RIP-Comm/clementine/assets/5256712/acad665d-cfac-4ec7-9333-d7eb4939fdd1)
![image](https://github.com/RIP-Comm/clementine/assets/5256712/9a75a718-f615-40c4-9e59-2f4297a1af9a)

Ok it works, nice :D

Reply:
I'm not quite sure that would work the same, as I need to keep the sign going from 32 to 64 bits, perhaps it might be the case to use the sign extension utility already available under `bitwise.rs`.
 
Simply doing `(rs_operand_value as i32) as i64` does the trick, but you guys tell me :)

Reply:
> I'm not quite sure that would work the same, as I need to preserve the sign value going from 32 to 64 bits, perhaps it might be the case to use the sign extension utility already available under `bitwise.rs`.
> 
> Simply doing `(rs_operand_value as i32) as i64` does the trick, but you guys tell me :)

I think what you suggested (`(rs_operand_value as i32) as i64`) is what I like most but given that we can't really avoid the double cast you can choose whatever you like most

Reply:
In fact `to_be -> from_be` is a bit silly, I probably went down this way because of the implementations of other opcodes that we've removed